feat: Schedule Return Dish Notifications: #125

MSNanda515 opened the pull request:
**Description**
PR for issue #125 
Schedule notifications for returning dishes

**Required Items**

- [x] Title matches [../CONTRIBUTING.md](contributing guidelines)
- [ ] Linked Issue: #XYZ
- [ ] Code Reviewed
- [x] Manually tested
- Style
  - [ ] No Lint Errors
  - [ ] Quality Comments
- Tests
  - [ ] Unit Tests Added
  - [ ] CI/CD passes
- UI
  - [ ] Picture of before/after UI (if applicable)

MSNanda515 commented:
@MacMalainey @Afraz0Khan 
Made the following changes to the schema:
- Added the email field to the user documents
- Added a new collection called config. It contains a document "admin" that further contains the field required such as "emailNotificationDelay"

MSNanda515 commented:
Removed all the secrets that were originally included.
Also, We will need an email account to complete the email notification functionality.

Afraz0Khan commented:
Looks good. I'll contact Mac to check with the email placeholder and see what we can do.

MacMalainey commented:
Please fix merge errors

> Looks good. I'll contact Mac to check with the email placeholder and see what we can do.

The user's email should be accessible using the Firebase Admin SDK, since transactions point to the UUID for the user. Please use that instead of a placeholder.

Unless I am misunderstanding what the placeholder email is for

MSNanda515 commented:
> Please fix merge errors
> 
> > Looks good. I'll contact Mac to check with the email placeholder and see what we can do.
> 
> The user's email should be accessible using the Firebase Admin SDK, since transactions point to the UUID for the user. Please use that instead of a placeholder.
> 
> Unless I am misunderstanding what the placeholder email is for

When you say UUID for the user, do you mean the document id will be the same as the CCID for the users, or do you mean that the user model will have a field containing the email (right now, there is no email field in the user's schema).

MSNanda515 commented:
Added code to get the user's model from Firebase Auth instead of Firestore. I also removed the email entry (I added before) from the User's schema in the software design document.

MacMalainey commented on server/controllers/scheduledJobsController.js:
```diff
@@ -0,0 +1,37 @@
+const scheduledJobsController = {
+
+    // Schedules a job to send the reminder to return the borrowed dish
+    dishCheckoutReturnReminder: async function (transId) {
```
Sorry, would we be able to do this as a batch job?

I.e. every Friday at 2:00pm check for transactions that are minimum N-days old and send an email to the users for them all at once (that way we can batch multiple missing dishes into one email together)

MSNanda515 replied:
Updated it to a batch job. The config needs to be in cron syntax for it to be valid.

MacMalainey commented on server/routes/scheduledJobs.js:
```diff
@@ -0,0 +1,18 @@
+
+// Schedules the reminder to return the dish
+// Requires transId, userId, dishId as query params
+router.get('/dishCheckout', async (req, res, next) => {
```
We shouldn't be triggering the reminder, we should be running a job every N-days to run a query on all outdated transactions

MacMalainey commented on server/services/emailService.js:
```diff
@@ -0,0 +1,33 @@
+import nodemailer from "nodemailer"
```
Should use the firebase email service

MacMalainey replied:
Will also reach out to DishZero regarding aquiring a more official email, but that can occur later

MSNanda515 replied:
Removed nodemailer. Will use the firebase email service once obtain details regarding the sender's email address.
